src/xdist/scheduler/customgroup.py

```diff
@@ -0,0 +1,348 @@
+
+    def add_node_collection(
+        self, node: WorkerController, collection: Sequence[str]
+    ) -> None:
+        """Add the collected test items from a node.
+
+        The collection is stored in the ``.node2collection`` map.
+        Called by the ``DSession.worker_collectionfinish`` hook.
+        """
+        assert node in self.node2pending
+        if self.collection_is_completed:
+            # A new node has been added later, perhaps an original one died.
+            # .schedule() should have
+            # been called by now
+            assert self.collection
+            if collection != self.collection:
+                other_node = next(iter(self.node2collection.keys()))
+                msg = report_collection_diff(
+                    self.collection, collection, other_node.gateway.id, node.gateway.id
+                )
+                self.log(msg)
+                return
+        self.node2collection[node] = list(collection)
+
+    def mark_test_complete(
+        self, node: WorkerController, item_index: int, duration: float = 0
+    ) -> None:
+        """Mark test item as completed by node.
+
+        The duration it took to execute the item is used as a hint to
+        the scheduler.
+
+        This is called by the ``DSession.worker_testreport`` hook.
+        """
+        self.node2pending[node].remove(item_index)
+        self.check_schedule(node, duration=duration)
+
+    def mark_test_pending(self, item: str) -> None:
+
+        assert self.collection is not None
+        self.pending.insert(
+            0,
+            self.collection.index(item),
+        )
+        for node in self.node2pending:
+            self.check_schedule(node)
+
+    def remove_pending_tests_from_node(
+        self,
+        node: WorkerController,
+        indices: Sequence[int],
+    ) -> None:
+        raise NotImplementedError()
+
+    def check_schedule(self, node: WorkerController, duration: float = 0, from_dsession=False) -> None:
+        """Maybe schedule new items on the node.
+
+        If there are any globally pending nodes left then this will
+        check if the given node should be given any more tests.  The
+        ``duration`` of the last test is optionally used as a
+        heuristic to influence how many tests the node is assigned.
+        """
+        if node.shutting_down:
+            self.terminal.write_line(f"{node.workerinput['workerid']} is already shutting down")
+            return
+
+        if self.pending:
+            any_working = False
+            for node in self.nodes:
+                if len(self.node2pending[node]) not in [0, 1]:
+                    any_working = True
+
+            if not any_working and from_dsession:
+                if self.pending_groups:
+                    dist_group_key = self.pending_groups.pop(0)
+                    dist_group = self.dist_groups[dist_group_key]
+                    nodes = cycle(self.nodes[0:dist_group['group_workers']])
```

Comment:
I left the check schedule signature unchanged, but we don't do anything if any node is working, so it is essentially node agnostic

Reply:
I apologize but I don't understand. `check_schedule` returns early if the node is checking down.
So there's the possibility that `node[0]` is shutting down while others are running (or viceversa) - i.e. this cares about which node is called on.

Reply:
Yes but if node[0] is shutting down we won't be scheduling anyways. We can remove that check though, I just didn't want to change the method signature, but we could also change and not pass in a node at all

Reply:
I'm more worried about the opposite: `node[0]` is not shutting down, so it jumpts to more scheduling, while the others are still running tests.
Also, since regular `schedule` is supposed to be `check_schedule` for all nodes after its first call, might it not be clearer to use that one?

Reply:
We check that they are all idle, so we wouldn't be rescheduling. We can remove the usage of node entirely.

I'm unsure what you mean, I think check_schedule is more clear rather than continuously calling schedule, but if behavior is the same either way I don't think it matters